packages/store/src/patterns/interface-tools.js

```diff
@@ -251,3 +251,145 @@ const emptyRecord = harden({});
  * @returns {{}}
  */
 export const initEmpty = () => emptyRecord;
+
+/**
+ * @template [S = any]
+ * @template [T = any]
+ * @typedef {object} Context
+ * @property {S} state
+ * @property {T} self
+ */
+
+/**
+ * @template A,S,T
+ * @param {string} tag
+ * @param {any} interfaceGuard
+ * @param {(...args: A[]) => S} init
+ * @param {T} methods
+ * @param {object} [options]
+ * @returns {(...args: A[]) => (T & RemotableBrand<{}, T>)}
+ */
+export const defineHeapFarClass = (
+  tag,
+  interfaceGuard,
+  init,
+  methods,
+  options = undefined,
+) => {
+  /** @type {WeakMap<T,Context<S,T>} */
+  const contextMap = new WeakMap();
+  const prototype = defendPrototype(
+    tag,
+    contextMap,
+    methods,
+    true,
+    interfaceGuard,
+  );
+  const makeInstance = (...args) => {
+    // Be careful not to freeze the state record
+    const state = seal(init(...args));
+    const self = harden({ __proto__: prototype });
+    // Be careful not to freeze the state record
+    /** @type {Context<S,T>} */
+    const context = freeze({ state, self });
```

Comment:
Doesn't this freeze the `state`?

Reply:
It does not. `harden` would because it applies `freeze` transitively. `freeze` itself is one level, which is why we use it here --- in order to avoid freezing the state.